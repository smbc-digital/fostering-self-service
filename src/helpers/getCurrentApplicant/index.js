import { Applicant, SECOND_APPLICANT_ROUTE } from 'constants'

const getCurrentApplicant = match => {
    if (match.params && match.params[0] === SECOND_APPLICANT_ROUTE) {
        return Applicant.SecondApplicant
    }

    return Applicant.FirstApplicant
}

export default getCurrentApplicant