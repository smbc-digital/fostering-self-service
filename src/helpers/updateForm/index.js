import { fetchWithTimeout } from '../index'

const TaskStatus = {
    None: 0,
    Completed: 1,
    NotCompleted: 2,
    CantStart: 3
}

const StageName = {
    HomeVisit: 0,
    Application: 1
}

const HomeVisitFormName = {
    TellUsAboutYourself: 0,
    YourEmploymentDetails: 1,
    LanguagesSpokenInYourHome: 2,
    YourPartnership: 3,
    YourFosteringHistory: 4,
    YourHealth: 5,
    TellUsAboutYourInterestInFostering: 6,
    YourHousehold: 7,
    ChildrenLivingAwayFromYourHome: 8
}

const ApplicationFormName = {
    References: 0,
    GpDetails: 1,
    AddressHistory: 2
}

const getHomeVisitUpdateEndpoint = form => {
    switch (form) {
        case HomeVisitFormName.TellUsAboutYourself:
            return '/fostering/home-visit/about-yourself'
        case HomeVisitFormName.YourEmploymentDetails:
            return '/fostering/home-visit/your-employment-details'
        case HomeVisitFormName.LanguagesSpokenInYourHome:
            return '/fostering/home-visit/languages-spoken-in-your-home'
        case HomeVisitFormName.YourFosteringHistory:
            return '/fostering/home-visit/your-fostering-history'
        case HomeVisitFormName.YourPartnership:
            return '/fostering/home-visit/partnership-status'
        case HomeVisitFormName.TellUsAboutYourInterestInFostering:
            return '/fostering/home-visit/interest-in-fostering'
        case HomeVisitFormName.YourHealth:
            return '/fostering/home-visit/about-your-health'
        case HomeVisitFormName.YourHousehold:
            return '/fostering/home-visit/household'
        case HomeVisitFormName.ChildrenLivingAwayFromYourHome:
            return '/fostering/home-visit/children-living-away-from-home'
        default:
            throw new Error('No matching endpoint for given form.')
    }
}

const getApplicationUpdateEndpoint = form => {
    switch (form) {
        case ApplicationFormName.References:
            return '/fostering/application/references'
        case ApplicationFormName.GpDetails:
            return '/fostering/application/gp-details'
        default:
            throw new Error('No matching endpoint for given form.')
    }
}

const getFormStageEndpoint = stage => {
    switch (stage) {
        case StageName.HomeVisit: 
            return '/fostering/home-visit'
        case StageName.Application:
            return '/fostering/application'
        default:
            throw new Error('No matching endpoint for given stage.')
    }
}

const reduceProperties = object => Object.keys(object).reduce((acc, property) => {
    return {
        ...acc,
        [property]: object[property].value
    }
}, {})

const updateForm = async (endpoint, formData) => {
    const parsedFormData = parseFormData(formData)

    const response = await fetchWithTimeout(endpoint, {
        method: 'PATCH',
        credentials: 'include',
        body: JSON.stringify(parsedFormData),
        headers: {
            'Accept': 'application/json',
            'Content-Type': 'application/json'
        }
    }, 30000)

    if (!response.ok) {
        throw Error(response.statusText)
    }

    return await response.json()
}

const updateFormStatus = ({ form, stage, currentStatus, setStatus }) => {
    if (currentStatus === TaskStatus.None) {
        fetchWithTimeout(`${getFormStageEndpoint(stage)}/status`,
            {
                method: 'PATCH',
                credentials: 'include',
                body: JSON.stringify({
                    status: TaskStatus.NotCompleted,
                    form
                }),
                headers: {
                    'Accept': 'application/json',
                    'Content-Type': 'application/json'
                }
            }, 30000)

        setStatus(TaskStatus.NotCompleted)
    }
}

const parseFormData = ({ firstApplicant, secondApplicant, familyReference, firstPersonalReference, secondPersonalReference, ...formData }) => {
    let parsedObject = reduceProperties(formData)

    if (firstApplicant) {
        parsedObject.firstApplicant = reduceProperties(firstApplicant)
    }

    if (secondApplicant) {
        parsedObject.secondApplicant = reduceProperties(secondApplicant)
    }

    if (familyReference) {
        parsedObject.familyReference = reduceProperties(familyReference)
        parsedObject.familyReference.address = familyReference.address.value
    }

    if (firstPersonalReference) {
        parsedObject.firstPersonalReference = reduceProperties(firstPersonalReference)
        parsedObject.firstPersonalReference.address = firstPersonalReference.address.value
    }

    if (secondPersonalReference) {
        parsedObject.secondPersonalReference = reduceProperties(secondPersonalReference)
        parsedObject.secondPersonalReference.address = secondPersonalReference.address.value
    }

    return parsedObject
}

const updateApplicationForm = async (form, formData) => {
    const endpoint = getApplicationUpdateEndpoint(form)

    return await updateForm(endpoint, formData)
}

const updateHomeVisitForm = async (form, formData) => {
    const endpoint = getHomeVisitUpdateEndpoint(form)

    return await updateForm(endpoint, formData)
}

export {
    TaskStatus,
    StageName,
    HomeVisitFormName,
    ApplicationFormName,
    updateFormStatus,
    parseFormData,
    updateApplicationForm,
    updateHomeVisitForm
}