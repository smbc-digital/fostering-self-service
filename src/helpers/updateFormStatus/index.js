export const TaskStatus = {
    None: 0,
    Completed: 1,
    NotCompleted: 2,
    CantStart: 3
}

export const FormName =
{
    TellUsAboutYourself: 0,
    YourEmploymentDetails: 1,
    LanguageSpokenInYourHome: 2,
    YourPartnership: 3,
    YourFosteringHistory: 4,
    YourHealth: 5,
    TellUsAboutYourInterestInFostering: 6,
    YourHousehold: 7,
    ChildrenLivingAwayFromYourHome: 7
}

export default function (form, currentStatus, setStatus) {
    if (currentStatus === TaskStatus.None) {
        fetch('/front-end-endpoint-update-status',
            {
                method: 'PATCH',
                credentials: 'include',
                body: JSON.stringify({
                    status: TaskStatus.NotCompleted,
                    form
                })
            })

        setStatus(TaskStatus.NotCompleted)
    }
}