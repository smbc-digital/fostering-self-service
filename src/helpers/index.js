import fetchWithTimeout from './fetchWithTimeout'
import { getPageRoute } from './pagehelper'
import getCurrentApplicant from './getCurrentApplicant'
import { updateFormStatus, updateForm, FormName, TaskStatus } from './updateForm'

export {
    fetchWithTimeout,
    getPageRoute,
    getCurrentApplicant,
    updateFormStatus,
    updateForm,
    FormName,
    TaskStatus
}