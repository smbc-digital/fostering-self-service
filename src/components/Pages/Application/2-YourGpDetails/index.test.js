import { React, mount, useContextMock, renderer } from 'helpers/SetupTest'
import YourGpDetails from './index'
import { TaskStatus } from 'helpers'
import * as helpers from 'helpers'

describe('YourGpDetails', () => {
    
    const onChangeStatusMock = jest.fn()
    const onChangeApplicantMock = jest.fn()

    beforeEach(() => {
        jest.resetAllMocks()
        useContextMock.mockReturnValue({
            statuses: {
                gpDetailsStatus: TaskStatus.Completed
            },
            onChangeStatus: onChangeStatusMock,
            onChangeApplicant: onChangeApplicantMock,
            firstApplicant: {
                nameOfGp: {
                    value: 'test',
                    isValid: true
                },
                nameOfGpPractice: {
                    value: 'test',
                    isValid: true
                },
                gpPhoneNumber: {
                    value: '01234567890',
                    isValid: true
                },
                gpAddress: {
                    value: {
                        addressLine1: 'line 1',
                        addressLine2: 'line 2',
                        town: 'town',
                        postcode: 'sk13xe'
                    },
                    isValid: true
                }
            },
            secondApplicant: null
        })
    })

    it('should update form status', () => {
        // Arrange
        useContextMock.mockReturnValue({
            statuses: {
                gpDetailsStatus: TaskStatus.None
            },
            onChangeStatus: onChangeStatusMock,
            onChangeApplicant: jest.fn(),
            firstApplicant: {
                nameOfGp: {
                    value: 'test',
                    isValid: true
                },
                nameOfGpPractice: {
                    value: 'test',
                    isValid: true
                },
                gpPhoneNumber: {
                    value: '01234567890',
                    isValid: true
                },
                gpAddress: {
                    value: {
                        addressLine1: 'line 1',
                        addressLine2: 'line 2',
                        town: 'town',
                        postcode: 'sk13xe'
                    },
                    isValid: true
                }
            },
            secondApplicant: null
        })
        const match = {
            params: []
        }

        helpers.updateFormStatus = jest.fn().mockImplementation((form, status, setStatus) => {
            setStatus(status)
        })

        // Act
        mount(<YourGpDetails match={match}/>)

        // Assert
        expect(onChangeStatusMock).toHaveBeenCalledWith('gpDetailsStatus', TaskStatus.None)
    })

    it('should push to next page on submit, when second applicant', () => {
        // Arrange
        useContextMock.mockReturnValue({
            statuses: {
                gpDetailsStatus: TaskStatus.None
            },
            onChangeStatus: onChangeStatusMock,
            onChangeApplicant: jest.fn(),
            firstApplicant: {
                nameOfGp: {
                    value: 'test',
                    isValid: true
                },
                nameOfGpPractice: {
                    value: 'test',
                    isValid: true
                },
                gpPhoneNumber: {
                    value: '01234567890',
                    isValid: true
                },
                gpAddress: {
                    value: {
                        addressLine1: 'line 1',
                        addressLine2: 'line 2',
                        town: 'town',
                        postcode: 'sk13xe'
                    },
                    isValid: true
                }
            },
            secondApplicant: {
                nameOfGp: {
                    value: 'test',
                    isValid: true
                },
                nameOfGpPractice: {
                    value: 'test',
                    isValid: true
                },
                gpPhoneNumber: {
                    value: '01234567890',
                    isValid: true
                },
                gpAddress: {
                    value: {
                        addressLine1: 'line 1',
                        addressLine2: 'line 2',
                        town: 'town',
                        postcode: 'sk13xe'
                    },
                    isValid: true
                }
            }
        })
        const history = {
            push: jest.fn()
        }
        const match = {
            params: []
        }

        helpers.updateForm = jest.fn().mockReturnValue(Promise.resolve(0)) 

        // Act
        const wrapper = mount(<YourGpDetails history={history} match={match} />)
        wrapper.find('form').simulate('submit')

        // Assert
        expect(history.push).toHaveBeenCalledWith(`${helpers.getPageRoute(22)}/second-applicant`)
    })

    it('should call updateForm on submit', async () => {
        //Arrange
        const history = {
            push: jest.fn()
        }
        const match = {
            params: []
        }
        helpers.updateForm = jest.fn().mockReturnValue(Promise.resolve(0)) 

        //Act
        const wrapper = mount(<YourGpDetails history={history} match={match} />)
        await wrapper.find('form').simulate('submit')
        await Promise.resolve()

        //Assert
        expect(helpers.updateForm).toHaveBeenCalled()
    })

    it('should call onChangeStatus on submit', async () => {
        // Arrange
        const history = {
            push: jest.fn()
        }
        const match = {
            params: []
        }
        helpers.updateForm = jest.fn().mockReturnValue(Promise.resolve(0)) 

        // Act
        const wrapper = mount(<YourGpDetails history={history} match={match} />)
        await wrapper.find('form').simulate('submit')
        await Promise.resolve()

        // Assert
        expect(onChangeStatusMock).toHaveBeenCalledWith('gpDetailsStatus', 0)
    })

    it('should push to the next page on submit', async () => {
        // Arrange
        const history = {
            push: jest.fn()
        }
        const match = {
            params: []
        }
        helpers.updateForm = jest.fn().mockReturnValue(Promise.resolve(0)) 

        // Act
        const wrapper = mount(<YourGpDetails history={history} match={match} />)
        await wrapper.find('form').simulate('submit')
        await Promise.resolve()

        // Assert
        expect(history.push).toHaveBeenCalledWith(helpers.getPageRoute(1))
    })

    it('should push to error page on submit', async () => {
        // Arrange
        const history = {
            push: jest.fn()
        }
        const match = {
            params: []
        }
        helpers.updateForm = jest.fn().mockImplementation(() => { throw new Error() }) 
        
        // Act
        const wrapper = mount(<YourGpDetails history={history} match={match} />)
        wrapper.find('form').simulate('submit')

        // Assert
        expect(history.push).toHaveBeenCalledWith('/error')
    })

    it('should call updateForm on save and go back click', () => {
        // Arrange
        const history = {
            push: jest.fn()
        }
        const match = {
            params: []
        }
        helpers.updateForm = jest.fn()

        // Act
        const wrapper = mount(<YourGpDetails history={history} match={match} />)
        wrapper.find('button').at(1).simulate('click')

        // Assert
        expect(helpers.updateForm).toHaveBeenCalled()
    })

    it('should push to the start page on save and go back click', async () => {
        // Arrange
        const history = {
            push: jest.fn()
        }
        const match = {
            params: []
        }
        helpers.updateForm = jest.fn().mockReturnValue(Promise.resolve(0)) 

        // Act
        const wrapper = mount(<YourGpDetails history={history} match={match} />)
        await wrapper.find('button').at(1).simulate('click')
        await Promise.resolve()

        // Assert
        expect(history.push).toHaveBeenCalledWith(helpers.getPageRoute(1))
    })

    it('should set isLoading on submit', () => {
        // Arrange
        const history = {
            push: jest.fn()
        }
        const match = {
            params: []
        }
        helpers.updateForm = jest.fn()

        // Act
        const wrapper = mount(<YourGpDetails history={history} match={match} />)
        wrapper.find('form').simulate('submit')

        // Assert
        expect(wrapper.find('SubmitButton').first().props().isLoading).toBe(true)
    })

    it('should call onChangeApplicant', () => {
        // Arrange
        const history = {
            push: jest.fn()
        }
        const match = {
            params: []
        }

        // Act
        const wrapper = mount(<YourGpDetails history={history} match={match} />)
        wrapper.find('input').first().simulate('change')

        // Assert
        expect(onChangeApplicantMock).toHaveBeenCalled()
    })

    describe('snapshot', () => {
        
        it('should render correctly', () => {
            const tree = renderer
                .create(<YourGpDetails history={{}} match={{ params: [] }} />)
                .toJSON()

            expect(tree).toMatchSnapshot()
        })
    })
})