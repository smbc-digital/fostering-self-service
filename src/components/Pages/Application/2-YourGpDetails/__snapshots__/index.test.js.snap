// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`YourGpDetails snapshot should render correctly 1`] = `
Array [
  <h1>
    Your fostering journey
  </h1>,
  <h2>
    Your GP details
  </h2>,
  <div>
    <section
      className="alert-information-form"
    >
      <span
        className="fa fa-info-circle"
      />
      <p
        dangerouslySetInnerHTML={
          Object {
            "__html": "We need to check that you’re fit and healthy enough to look after a child. We’ll pay for you to have a medical assessment from your GP and our medical advisor will then talk to your social worker about your fitness to foster.",
          }
        }
      />
    </section>
  </div>,
  <form
    onSubmit={[Function]}
  >
    <label
      className={undefined}
      htmlFor="nameOfGp"
    >
      Name of GP
    </label>
    <input
      aria-label="Name of GP"
      aria-live="off"
      className=""
      id="nameOfGp"
      inputMode={undefined}
      maxLength={60}
      min={undefined}
      name="nameOfGp"
      onChange={[Function]}
      onKeyDown={[Function]}
      pattern={undefined}
      type="text"
      value="test"
    />
    <label
      className={undefined}
      htmlFor="nameOfGpPractice"
    >
      Name of GP practice
    </label>
    <input
      aria-label="Name of GP practice"
      aria-live="off"
      className=""
      id="nameOfGpPractice"
      inputMode={undefined}
      maxLength={60}
      min={undefined}
      name="nameOfGpPractice"
      onChange={[Function]}
      onKeyDown={[Function]}
      pattern={undefined}
      type="text"
      value="test"
    />
    <label
      className={undefined}
      htmlFor="gpPhoneNumber"
    >
      Phone number
    </label>
    <input
      aria-label="Phone number"
      aria-live="off"
      className=""
      id="gpPhoneNumber"
      inputMode={undefined}
      maxLength={11}
      min={undefined}
      name="gpPhoneNumber"
      onChange={[Function]}
      onKeyDown={[Function]}
      pattern={undefined}
      type="tel"
      value="01234567890"
    />
    <label
      className={undefined}
      htmlFor="addressLine1"
    >
      Address line 1
    </label>
    <input
      aria-label="Address line 1"
      aria-live="off"
      className=""
      id="addressLine1"
      inputMode={undefined}
      maxLength={undefined}
      min={undefined}
      name="addressLine1"
      onChange={[Function]}
      onKeyDown={[Function]}
      pattern={undefined}
      type="text"
      value="line 1"
    />
    <h3
      aria-hidden={true}
      className="optional"
    >
      Address line 2
    </h3>
    <p
      aria-hidden={true}
      className="optional"
    >
      (optional)
    </p>
    <label
      className="optional sr-only"
      htmlFor="addressLine2"
    >
      Address line 2 optional
    </label>
    <input
      aria-label={undefined}
      aria-live={undefined}
      className=""
      id="addressLine2"
      inputMode={undefined}
      maxLength={undefined}
      min={undefined}
      name="addressLine2"
      onChange={[Function]}
      onKeyDown={[Function]}
      pattern={undefined}
      type="text"
      value="line 2"
    />
    <label
      className={undefined}
      htmlFor="town"
    >
      Town
    </label>
    <input
      aria-label="Town"
      aria-live="off"
      className=""
      id="town"
      inputMode={undefined}
      maxLength={undefined}
      min={undefined}
      name="town"
      onChange={[Function]}
      onKeyDown={[Function]}
      pattern={undefined}
      type="text"
      value="town"
    />
    <label
      className={undefined}
      htmlFor="postcode"
    >
      Postcode
    </label>
    <input
      aria-label="Postcode"
      aria-live="off"
      className=""
      id="postcode"
      inputMode={undefined}
      maxLength={undefined}
      min={undefined}
      name="postcode"
      onChange={[Function]}
      onKeyDown={[Function]}
      pattern={undefined}
      type="text"
      value="sk13xe"
    />
    <p>
      <a
        href="#"
        onClick={[Function]}
      >
         Use address finder 
      </a>
    </p>
    <button
      className="button-primary"
      disabled={false}
      id={undefined}
      onClick={undefined}
      type="submit"
    >
      Save and next step
    </button>
    <p>
      <a
        href="#"
        onClick={[Function]}
      >
        Previous
      </a>
    </p>
    <button
      className="button-inverted left-chevron"
      disabled={false}
      id={undefined}
      onClick={[Function]}
      type="submit"
    >
      Save and go back to fostering area
    </button>
  </form>,
]
`;
